ament_cmake_python/cmake/ament_python_install_package.cmake

```diff
@@ -42,34 +47,90 @@ function(_ament_cmake_python_install_package package_name)
     set(ARG_PACKAGE_DIR "${CMAKE_CURRENT_LIST_DIR}/${ARG_PACKAGE_DIR}")
   endif()
 
+  if(NOT ARG_VERSION)
+    # Use package.xml version
+    if(NOT _AMENT_PACKAGE_NAME)
+      ament_package_xml()
+    endif()
+    set(ARG_VERSION "${${PROJECT_NAME}_VERSION}")
+  endif()
+
   if(NOT EXISTS "${ARG_PACKAGE_DIR}/__init__.py")
     message(FATAL_ERROR "ament_python_install_package() the Python package "
       "folder '${ARG_PACKAGE_DIR}' doesn't contain an '__init__.py' file")
   endif()
 
-  _ament_cmake_python_register_environment_hook()
-
-  if(NOT PYTHON_INSTALL_DIR)
-    message(FATAL_ERROR "ament_python_install_package() variable "
-      "'PYTHON_INSTALL_DIR' must not be empty")
+  if(NOT ARG_SETUP_CFG)
+    if(EXISTS "${CMAKE_CURRENT_LIST_DIR}/setup.cfg")
+      set(ARG_SETUP_CFG "${CMAKE_CURRENT_LIST_DIR}/setup.cfg")
+    endif()
+  elseif(NOT IS_ABSOLUTE "${ARG_SETUP_CFG}")
+    set(ARG_SETUP_CFG "${CMAKE_CURRENT_LIST_DIR}/${ARG_SETUP_CFG}")
   endif()
-  install(
-    DIRECTORY "${ARG_PACKAGE_DIR}/"
-    DESTINATION "${PYTHON_INSTALL_DIR}/${package_name}"
-    PATTERN "*.pyc" EXCLUDE
-    PATTERN "__pycache__" EXCLUDE
+
+  set(build_dir "${CMAKE_CURRENT_BINARY_DIR}/ament_cmake_python/${package_name}")
+  file(RELATIVE_PATH source_dir "${build_dir}" "${ARG_PACKAGE_DIR}")
+
+  string(CONFIGURE "\
+import os
+from setuptools import find_packages
+from setuptools import setup
+
+setup(
+  name='${package_name}',
+  version='${ARG_VERSION}',
+  packages=find_packages(
+    where=os.path.normpath('${source_dir}/..'),
+    include=('${package_name}', '${package_name}.*')),
+  package_dir={'${package_name}': '${source_dir}'},
+  package_data={'': ['*.*']}
```

Comment:
If this is mimicking the install rule copying the entire package dir, then it seems like the pattern should be `*` instead of `*.*` to include files without dots.

Might need something like `exclude_package_data={'': ['*.pyc', '__pycache__']}` to avoid copying any byte code files in the source directory like the `install()` rule's `EXCLUDE` patterns, though it seems like that might not be enough if there are byte code files in a subdirectory and `exclude_package_data` also doesn't support recursive glob patterns https://github.com/pypa/setuptools/issues/1806 . I don't have any ideas on how to fix it, but I'm fine with not fixing it.

Reply:
You're completely right. My implementation was biased by our use cases.

I think I got it right in https://github.com/ament/ament_cmake/pull/316/commits/7876939bd429a38f60d1b7339ef1285990349dfd and https://github.com/ament/ament_cmake/pull/316/commits/a2f1349dd42c3630116beb1c39da60e4998aab86. PTAL !